### PR TITLE
[Feature] Differentiable image histogram using kernel density estimation.

### PR DESCRIPTION
**Pull request by bsuleymanov**
### Description
Differentiable histogram estimation for (B, C, H, W)-shaped tensors, #790.

### Status
**Work in progress**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [x] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>

  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?

</details>

### CODE REVIEW

**Review by shijianjian — CHANGES_REQUESTED** [review]
Just some docs and tests. BTW, please try reduce the amount of assertions if possible.

**Comment by bsuleymanov** [comment]
> BTW, please try reduce the amount of assertions if possible.
New here, don't know conventions yet, tried to follow exceptions-code from neighbor functions. Could you tell me, how to decide whether to insert an assertion? I will try to reduce the amount of them.

File: kornia/enhance/histogram.py [hunk 1]
```diff
@@ -1,8 +1,9 @@
+import math
 from typing import Tuple
 
 import torch
 
-__all__ = ["histogram", "histogram2d"]
+__all__ = ["histogram", "histogram2d", "image_hist2d"]
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:6]: I'd advocate for full names `image_histogram2d`

File: kornia/enhance/histogram.py [hunk 2]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+def image_hist2d(
+    image: torch.Tensor,
+    min: float = 0.0,
+    max: float = 255.0,
+    n_bins: int = 256,
+    bandwidth: float = -1.0,
+    centers: torch.Tensor = torch.tensor([]),
+    return_pdf: bool = False,
+    kernel: str = "triangular",
+):
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:165]:
```suggestion
) -> Tuple[torch.Tensor, torch.Tensor]:
```

File: kornia/enhance/histogram.py [hunk 3]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+    """Estimate the histogram of the input image(s).
+
+    The calculation uses triangular kernel density estimation.
+
+    Args:
+        image: Input tensor to compute the histogram with shape
+        :math:`(H, W)`, :math:`(C, H, W)` or :math:`(B, C, H, W)`.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:172]:
```suggestion
          :math:`(H, W)`, :math:`(C, H, W)` or :math:`(B, C, H, W)`.
```

File: kornia/enhance/histogram.py [hunk 4]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        min: Lower end of the interval (inclusive).
+        max: Upper end of the interval (inclusive). Ignored when
+        :attr:`centers` is specified.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:175]:
```suggestion
          :attr:`centers` is specified.
```

File: kornia/enhance/histogram.py [hunk 5]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        n_bins: The number of histogram bins. Ignored when
+        :attr:`centers` is specified.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:177]:
```suggestion
          :attr:`centers` is specified.
```

File: kornia/enhance/histogram.py [hunk 6]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        bandwidth: Smoothing factor. If not specified or equal to -1,
+        bandwidth = (max - min) / n_bins.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:179]:
```suggestion
          :math:`(bandwidth = (max - min) / n_bins)`.
```

File: kornia/enhance/histogram.py [hunk 7]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        centers: Centers of the bins with shape :math:`(n_bins,)`.
+        If not specified or empty, it is calculated as centers of
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:181]:
```suggestion
          If not specified or empty, it is calculated as centers of
```

File: kornia/enhance/histogram.py [hunk 8]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        equal width bins of [min, max] range.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:182]:
```suggestion
          equal width bins of [min, max] range.
```

File: kornia/enhance/histogram.py [hunk 9]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        return_pdf: If True, also return probability densities for
+        each bin.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:184]:
```suggestion
          each bin.
```

File: kornia/enhance/histogram.py [hunk 10]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        kernel: kernel to perform kernel density estimation
+        (`triangular`, `gaussian`, `uniform`, `epanechnikov`).
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:186]:
```suggestion
          ``(`triangular`, `gaussian`, `uniform`, `epanechnikov`)``.
```

File: kornia/enhance/histogram.py [hunk 11]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+
+    Returns:
+        Computed histogram of shape :math:`(bins)`, :math:`(C, bins)`,
+        :math:`(B, C, bins)`.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:190]:
```suggestion
          :math:`(B, C, bins)`.
```

File: kornia/enhance/histogram.py [hunk 12]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        Computed probability densities of shape :math:`(bins)`, :math:`(C, bins)`,
+        :math:`(B, C, bins)`, if return_pdf is ``True``. Tensor of zeros with shape
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:192]:
```suggestion
          :math:`(B, C, bins)`, if return_pdf is ``True``. Tensor of zeros with shape
```

File: kornia/enhance/histogram.py [hunk 13]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        of the histogram otherwise.
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:193]:
```suggestion
          of the histogram otherwise.
```

File: kornia/enhance/histogram.py [hunk 14]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        centers = min + bandwidth * (torch.arange(n_bins, device=device).float() + 0.5)
+    centers = centers.reshape(-1, 1, 1, 1, 1)
+    u = abs(image.unsqueeze(0) - centers) / bandwidth
+    if kernel == "triangular":
+        mask = (u <= 1).float()
+        kernel_values = (1 - u) * mask
+    elif kernel == "gaussian":
+        kernel_values = torch.exp(-0.5 * u ** 2)
+    elif kernel == "uniform":
+        mask = (u <= 1).float()
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:244]:
```suggestion
        mask = (u <= 1).to(u.dtype)
```

File: kornia/enhance/histogram.py [hunk 15]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        batch_size, n_channels = 1, 1
+    else:
+        raise ValueError(f"Input values must be a tensor of the shape " f"BxCxHxW, CxHxW or HxW. Got {image.shape}.")
+
+    if bandwidth == -1.0:
+        bandwidth = (max - min) / n_bins
+    if centers.numel() == 0:
+        centers = min + bandwidth * (torch.arange(n_bins, device=device).float() + 0.5)
+    centers = centers.reshape(-1, 1, 1, 1, 1)
+    u = abs(image.unsqueeze(0) - centers) / bandwidth
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:237]: could you add a comment what's this variable. Naming `u` is not explicit enough

File: kornia/enhance/histogram.py [hunk 16]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        kernel_values = torch.ones_like(u, dtype=u.dtype, device=u.device) * mask
+    elif kernel == "epanechnikov":
+        mask = (u <= 1).float()
+        kernel_values = (1 - u ** 2) * mask
+    else:
+        raise ValueError(f"Kernel must be 'triangular', 'gaussian', " f"'uniform' or 'epanechnikov'. Got {kernel}.")
+
+    hist = torch.sum(kernel_values, dim=(-2, -1)).permute(1, 2, 0)
+    if return_pdf:
+        normalization = torch.sum(hist, dim=-1).unsqueeze(0) + 1e-10
```
Comment (edgarriba) [inline kornia/enhance/histogram.py:254]:
```suggestion
        normalization = torch.sum(hist, dim=-1).unsqueeze(0) + eps
```
and add to `eps` to the function signature

File: test/enhance/test_histogram.py [hunk 17]
```diff
@@ -6,103 +7,178 @@
...
 
-class TestHistogram:
+class TestImageHist2d:
+    fcn = kornia.enhance.image_hist2d
+
+    @pytest.mark.parametrize("kernel", ["triangular", "gaussian", "uniform", "epanechnikov"])
+    def test_shape(self, device, dtype, kernel):
+        input = torch.ones(32, 32, device=device, dtype=dtype)
+        hist, pdf = TestImageHist2d.fcn(input, 0.0, 1.0, 256, kernel=kernel)
+        assert hist.shape == (1, 1, 256) and pdf.shape == (1, 1, 256)
```
Comment (edgarriba) [inline test/enhance/test_histogram.py:17]: I would expect this to return `(256,)`

File: test/enhance/test_histogram.py [hunk 18]
```diff
@@ -6,103 +7,178 @@
...
+
+    @pytest.mark.parametrize("device", [("cuda"), ("cpu")])
+    @pytest.mark.parametrize("kernel", ["triangular", "gaussian", "uniform", "epanechnikov"])
+    def test_shape_channels(self, device, dtype, kernel):
+        input = torch.ones(3, 32, 32, device=device, dtype=dtype)
+        hist, pdf = TestImageHist2d.fcn(input, 0.0, 1.0, 256, kernel=kernel)
+        assert hist.shape == (1, 3, 256) and pdf.shape == (1, 3, 256)
```
Comment (edgarriba) [inline test/enhance/test_histogram.py:24]: I would expect this to return `(3, 256)`

File: test/enhance/test_histogram.py [hunk 19]
```diff
@@ -6,103 +7,178 @@
...
+class TestImageHist2d:
+    fcn = kornia.enhance.image_hist2d
+
+    @pytest.mark.parametrize("kernel", ["triangular", "gaussian", "uniform", "epanechnikov"])
+    def test_shape(self, device, dtype, kernel):
+        input = torch.ones(32, 32, device=device, dtype=dtype)
+        hist, pdf = TestImageHist2d.fcn(input, 0.0, 1.0, 256, kernel=kernel)
+        assert hist.shape == (1, 1, 256) and pdf.shape == (1, 1, 256)
+
+    @pytest.mark.parametrize("device", [("cuda"), ("cpu")])
```
Comment (edgarriba) [inline test/enhance/test_histogram.py:19]:
```suggestion
```
you don't need to add this line. It's done automatically during the testing setup

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: test/enhance/test_histogram.py (hunk 17 above)
Comment (bsuleymanov) [reply test/enhance/test_histogram.py]: Thanks, I've fixed it

File: kornia/enhance/histogram.py [hunk 20]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        Computed histogram of shape :math:`(bins)`, :math:`(C, bins)`,
+        :math:`(B, C, bins)`.
+        Computed probability densities of shape :math:`(bins)`, :math:`(C, bins)`,
+        :math:`(B, C, bins)`, if return_pdf is ``True``. Tensor of zeros with shape
+        of the histogram otherwise.
+    """
+    if not isinstance(image, torch.Tensor):
+        raise TypeError(f"Input image type is not a torch.Tensor. Got {type(image)}.")
+
+    if centers is not None and not isinstance(centers, torch.Tensor):
```
[the remark this answers is not in the document]
Comment (bsuleymanov) [reply kornia/enhance/histogram.py:199]: Fixed

File: kornia/enhance/histogram.py [hunk 21]
```diff
@@ -150,3 +151,107 @@ def histogram2d(
...
+        n_channels, height, width = image.size()
+    elif image.dim() == 2:
+        height, width = image.size()
+        batch_size, n_channels = 1, 1
+    else:
+        raise ValueError(f"Input values must be a tensor of the shape " f"BxCxHxW, CxHxW or HxW. Got {image.shape}.")
+
+    if bandwidth == -1.0:
+        bandwidth = (max - min) / n_bins
+    if centers.numel() == 0:
```
[the remark this answers is not in the document]
Comment (bsuleymanov) [reply kornia/enhance/histogram.py]: Updated

File: kornia/enhance/histogram.py (hunk 15 above)
Comment (bsuleymanov) [reply kornia/enhance/histogram.py]: Tried to explain

**Comment by edgarriba** [comment]
@bsuleymanov are the tests passing in local for you ?

**Comment by bsuleymanov** [comment]
> @bsuleymanov are the tests passing in local for you ?

yes, 28 tests from TestImageHistogram2d are passed

**Comment by edgarriba** [comment]
Ok, it’s weird- none of the tests pass in the CI. Are you in our slack ? We could discuss there

**Comment by bsuleymanov** [comment]
> Ok, it’s weird- none of the tests pass in the CI. Are you in our slack ? We could discuss there

My fault, I forgot to push commits related to the tests :)

**Comment by bsuleymanov** [comment]
> Ok, it’s weird- none of the tests pass in the CI. Are you in our slack ? We could discuss there

jit tests still fail (locally they are passed), maybe because of returned tuple. Yes, I'm in your slack, @Bulat Suleymanov

File: test/enhance/test_histogram.py [hunk 22]
```diff
@@ -6,103 +7,172 @@
...
+    @pytest.mark.parametrize("kernel", ["triangular", "gaussian", "uniform", "epanechnikov"])
+    def test_gradcheck(self, device, dtype, kernel):
+        input = torch.ones(32, 32, device=device, dtype=dtype)
+        input = utils.tensor_to_gradcheck_var(input)  # to var
+        centers = torch.linspace(0, 255, 8, device=device, dtype=dtype)
+        centers = utils.tensor_to_gradcheck_var(centers)
+        assert gradcheck(
+            TestImageHistogram2d.fcn, (input, 0.0, 255.0, 256, None, centers, True, kernel), raise_exception=True
+        )
+
```
Comment (edgarriba) [inline test/enhance/test_histogram.py:40]:
```suggestion
    @pytest.mark.skipif(version.parse(torch.__version__) < version.parse("1.9"),
                                    reason="Tuple cannot be jitted with PyTorch < v1.9")
```

File: test/enhance/test_histogram.py [hunk 23]
```diff
@@ -1,3 +1,4 @@
+import pytest
```
Comment (edgarriba) [inline test/enhance/test_histogram.py:1]:
```suggestion
from packaging import version
import pytest
```

**Comment by edgarriba** [comment]
@bsuleymanov would you like to contribute with a tutorial showing the usage of this functionality ?

**Comment by bsuleymanov** [comment]
> @bsuleymanov would you like to contribute with a tutorial showing the usage of this functionality ?

Yes, I would

**Comment by edgarriba** [comment]
@bsuleymanov awesome - the way we make tutorials now is through jupyter notebooks in this repository https://github.com/kornia/tutorials. Feel free to send a PR there.

**Comment by bsuleymanov** [comment]
@edgarriba Okay, I will start next week
